### PR TITLE
Move from nocrypto to mirage-crypto

### PR DESCRIPTION
**Pull request by ulrikstrid**
This will only work in ocaml 4.07+, if that's fine I think this is a good way forward.

### CODE REVIEW

**Comment by talex5** [comment]
Looks good to me! Would be good to get a release in opam-repository to avoid pulling in the old nocrypto library.

**Comment by seliopou** [comment]
I think that all the OPAM files now need a lower bound on the version of session to ensure that they're all picking up the same rng initialization.

**Comment by ulrikstrid** [comment]
Would it be easiest to do a release and fix the reverse dependencies from there or hunt then when beforehand?
I'm not sure how the OCaml ecosystem works exactly.

**Comment by seliopou** [comment]
Fixing reverse dependencies is not necessary. The packages in this library that depend on session from this point on have to depend on a later version than the current, which is `session.0.4.1`. Any released packages from this library that depend on session have to depend on an earlier or current version, which is the same `session.0.4.1`.

File: backends/postgresql/session_postgresql.ml [hunk 1]
```diff
@@ -65,7 +65,7 @@ let connect ?host ?hostaddr ?port ?dbname ?user ?password ?options ?tty ?require
   new connection ?host ?hostaddr ?port ?dbname ?user ?password ?options ?tty ?requiressl ?conninfo ?startonly ()
 
 let gensym () =
-  Cstruct.to_string Nocrypto.(Base64.encode (Rng.generate 30))
+  Base64.encode_exn (Cstruct.to_string (Mirage_crypto_rng.generate 30))
```
Comment (hannesm) [inline backends/postgresql/session_postgresql.ml]:
```suggestion
  Base64.encode_string (Cstruct.to_string (Mirage_crypto_rng.generate 30))
```

File: backends/redis/lwt/session_redis_lwt.ml [hunk 2]
```diff
@@ -52,7 +52,7 @@ let _default_period =
   Int64.of_int (60 * 60 * 24 * 7)
 
 let gensym () =
-  Cstruct.to_string Nocrypto.(Base64.encode (Rng.generate 30))
+  Base64.encode_exn (Cstruct.to_string (Mirage_crypto_rng.generate 30))
```
Comment (hannesm) [inline backends/redis/lwt/session_redis_lwt.ml]:
```suggestion
  Base64.encode_string (Cstruct.to_string (Mirage_crypto_rng.generate 30))
```

File: backends/redis/lwt/dune [hunk 3]
```diff
@@ -1,4 +1,4 @@
 (library
  (name session_redis_lwt)
  (public_name session-redis-lwt)
- (libraries lwt nocrypto redis-lwt result session))
+ (libraries lwt mirage-crypto redis-lwt result session))
```
Comment (hannesm) [inline backends/redis/lwt/dune]:
```suggestion
 (libraries lwt mirage-crypto-rng base64 redis-lwt result session))
```

File: backends/postgresql/dune [hunk 4]
```diff
@@ -1,4 +1,4 @@
 (library
  (name session_postgresql)
  (public_name session-postgresql)
- (libraries nocrypto postgresql result session threads))
+ (libraries mirage-crypto postgresql result session threads))
```
Comment (hannesm) [inline backends/postgresql/dune]:
```suggestion
 (libraries mirage-crypto-rng base64 postgresql result session threads))
```

File: lib/memory.ml [hunk 5]
```diff
@@ -44,7 +44,7 @@ type t =
   ; mutable default_period : period }
 
 let gensym () =
-  Cstruct.to_string Nocrypto.(Base64.encode (Rng.generate 30))
+  Base64.encode_exn (Cstruct.to_string (Mirage_crypto_rng.generate 30))
```
Comment (hannesm) [inline lib/memory.ml]:
```suggestion
  Base64.encode_string (Cstruct.to_string (Mirage_crypto_rng.generate 30))
```

File: session.opam [hunk 6]
```diff
@@ -8,7 +8,9 @@ doc: "https://inhabitedtype.github.io/ocaml-session/"
 depends: [
   "ocaml" {>= "4.03.0"}
```
Comment (hannesm) [inline session.opam]: this should then be 4.07.0 as well?

File: session.opam [hunk 7]
```diff
@@ -8,7 +8,9 @@ doc: "https://inhabitedtype.github.io/ocaml-session/"
...
   "dune" {build & >= "1.0"}
-  "nocrypto"
+  "mirage-crypto"
+  "mirage-crypto-rng"
+  "base64"
   "result"
```
Comment (hannesm) [inline session.opam]: this seems to be superfluous now (since 4.03.0 the `type ('a, 'b) result` is part of Stdlib/Pervasives).

**Comment by ulrikstrid** [comment]
I have now addressed the version constraints @seliopou and @hannesm comments. I did the version constraints as suggested by @kit-ty-kate in your opam PR, let me know if you want me to change that.

**Comment by seliopou** [comment]
Adding an exact version constraint is not necessary. Please change it to a non-inclusive lower bound for the latest release in OPAM.

**Comment by ulrikstrid** [comment]
Okay I'll do it tomorrow

**Comment by ulrikstrid** [comment]
I hope `{>= "0.4.2"}` was correct here. Unless you plan on publishing this as 0.5.0 since this is breaking, then we could just set it to `{>= "0.5.0"}` so you're free to publish patches to the old versions if that is needed

**Comment by seliopou** [comment]
Thanks!

**Pull request merged**